[nrf noup] Added support for multi-variant application OTA DFU

markaj-nordic opened the pull request:
* Implemented dedicated flash write handlers for a Thread/WiFi switchable app.
* A script for generating multi variant OTA image has been added in sdk-nrf.

Signed-off-by: Marcin Kajor <marcin.kajor@nordicsemi.no>

markaj-nordic commented on src/platform/nrfconnect/OTAMultiImageDownloader.cpp:
```diff
@@ -0,0 +1,181 @@
+    return err;
+}
+
+// Implementation of stateless OTAMultiImageDownloaders class
```
Maybe this can be replaced with just a namespace.

Damian-Nordic commented on src/platform/nrfconnect/OTAImageProcessorImpl.cpp:
```diff
@@ -114,6 +144,10 @@ CHIP_ERROR OTAImageProcessorImpl::Apply()
     int err = dfu_target_schedule_update(-1);
 
+#ifdef CONFIG_THREAD_WIFI_SWITCHING
+    err = OTAMultiImageDownloaders::Apply();
```
Shouldn't it be rather called in `ConfirmCurrentImage`? That is, when we know that the image has been applied correctly and we already execute it successfully?

markaj-nordic replied:
Ahh sure, my bad.

Damian-Nordic commented on config/nrfconnect/chip-module/Kconfig.defaults:
```diff
@@ -173,6 +173,11 @@ config NFC_THREAD_CALLBACK
 config CHIP_OTA_REQUESTOR
     default y
 
+if THREAD_WIFI_SWITCHING
```
This is defined just in the lock sample, right? If so, it may be better to modify `DFU_MULTI_IMAGE_MAX_IMAGE_COUNT` there?

markaj-nordic replied:
Ok, I will move it to the lock sample.

Damian-Nordic commented on src/platform/nrfconnect/OTAMultiImageDownloader.cpp:
```diff
@@ -0,0 +1,181 @@
+
+int OTAMultiImageDownloader::Init(const device * flash, size_t offset, size_t size)
+{
+    mBuffer = (uint8_t *) k_malloc(kBufferSize);
```
I think using kernel heap may not be the best choice given how precious resource there is. We should better use `Platform::MemoryAlloc` or even `ScopedBuffer` for safer memory management.

markaj-nordic replied:
Ok, will do.

Damian-Nordic commented on src/platform/nrfconnect/OTAMultiImageDownloader.cpp:
```diff
@@ -0,0 +1,181 @@
+
+int OTAMultiImageDownloaders::Write(const uint8_t * chunk, size_t chunk_size)
+{
+    if (OTAMultiImageDownloaders::CurrentImageId() <= CurrentImageID::APP_1_CORE_NET)
```
I struggle to understand this condition. What if we're currently running a Wi-Fi variant?

markaj-nordic replied:
Yep, look like a bug. Will fix it.

Damian-Nordic reviewed (commented):
A few initial comments. I wonder if it wouldn't be cleaner if you created another `MultiVariantOTAImageProcessor` class that inherits from `OTAImageProcessorImpl`, overrides needed methods and keeps all the necessary state in itself. That way you could avoid mixing instance-based code in `OTAImageProcessorImpl` with `OTAMultiImageDownloaders` that uses global state. However, I admit we have plenty of code that uses global state right now, so I'm fine with the current approach, either :)

markaj-nordic commented:
> A few initial comments. I wonder if it wouldn't be cleaner if you created another `MultiVariantOTAImageProcessor` class that inherits from `OTAImageProcessorImpl`, overrides needed methods and keeps all the necessary state in itself. That way you could avoid mixing instance-based code in `OTAImageProcessorImpl` with `OTAMultiImageDownloaders` that uses global state. However, I admit we have plenty of code that uses global state right now, so I'm fine with the current approach, either :)

I get your point. Let me see if I can embrace this approach. For now I have addressed all comments regarding current implementation.

markaj-nordic commented:
> A few initial comments. I wonder if it wouldn't be cleaner if you created another `MultiVariantOTAImageProcessor` class that inherits from `OTAImageProcessorImpl`, overrides needed methods and keeps all the necessary state in itself. That way you could avoid mixing instance-based code in `OTAImageProcessorImpl` with `OTAMultiImageDownloaders` that uses global state. However, I admit we have plenty of code that uses global state right now, so I'm fine with the current approach, either :)

I get your point. Let me see if I can embrace this approach. For now I have addressed all comments regarding current implementation.

> > A few initial comments. I wonder if it wouldn't be cleaner if you created another `MultiVariantOTAImageProcessor` class that inherits from `OTAImageProcessorImpl`, overrides needed methods and keeps all the necessary state in itself. That way you could avoid mixing instance-based code in `OTAImageProcessorImpl` with `OTAMultiImageDownloaders` that uses global state. However, I admit we have plenty of code that uses global state right now, so I'm fine with the current approach, either :)
> 
> I get your point. Let me see if I can embrace this approach. For now I have addressed all comments regarding current implementation.

The problem I see with this approach is that we need to register dfu_multi_image callbacks which by definition must be implemented as stateless (i.e. we cannot capture any non static member of OTAImageProcessorImpl). Yet we still have to track the currently processed image ID to decide which external flash slots must be targeted for a given image chunk. So it looks like we cannot do without a globally scoped state (current image ID). It seems to me that it could be resolved by modification of dfu_multi_image API by either providing current image ID in all callbacks (as it is done only for open callback currently) or (maybe) by allowing a user to register multiple writers for given ID. Regardless of that we can still inherit from OTAImageProcessorImpl but I am not sure if this brings an added value (well it's a tradeoff between having more neat code/more files in the repo).

kkasperczyk-no commented on config/nrfconnect/chip-module/CMakeLists.txt:
```diff
@@ -222,6 +222,7 @@ chip_gn_arg_bool  ("chip_detail_logging"                    CONFIG_MATTER_LOG_LE
 chip_gn_arg_bool  ("chip_automation_logging"                "false")
 chip_gn_arg_bool  ("chip_malloc_sys_heap"                   CONFIG_CHIP_MALLOC_SYS_HEAP)
 chip_gn_arg_bool  ("chip_enable_wifi"                       CONFIG_WIFI_NRF700X)
+chip_gn_arg_bool  ("chip_multi_variant_ota"                 CONFIG_THREAD_WIFI_SWITCHING)
```
That's a minor, but maybe it would be cleaner to create a dedicated Kconfig? I would imagine it could be CONFIG_MULTI_VARIANT_OTA and then imply it in the CONFIG_THREAD_WIFI_SWITCHING.

markaj-nordic replied:
IMHO the clean solution would be simply to support OTA whenever the CHIP_OTA_REQUESTOR is set, regardless if we have single or multiple variant application. This is actually the current state. On the other hand I agree that mapping the CONFIG_THREAD_WIFI_SWITCHING into chip_multi_variant_ota does not look clean... but @kkasperczyk-no do you have any idea how to make it more neat without introducing another Kconfig (which preferably I would like to avoid)? I mean we may just build all OTA related files but that's rather a no-go considering our flash limitations.

kkasperczyk-no replied:
But why would you like to avoid it? I know it may seem we have already bunch of kconfig options, but the fact is that our platform is very complex and this is the most effective way of configuring our implementation to disable certain parts of a code. 
Honestly I thought this functionality will be used only by the switched sample, so we will somehow implement it in the sample code to avoid making our platform implementation even more complex with another variant of it. Nevertheless I understand it is much more convenient to do that in the platform. In such case I would recommend wrap it into new kconfig option to make some new functionality clearly separated from the single sample.

markaj-nordic commented:
After f2f discussion the decision is to move these implementation to nrfconnect SDK. Closing this PR.
